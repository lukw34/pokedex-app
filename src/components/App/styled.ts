import styled, { createGlobalStyle } from 'styled-components';


export const GlobalStyled = createGlobalStyle`
    body {
        overflow: hidden;
    }
`;

export const AppStyledBackground = styled.div`
    position: absolute;
    width: 100%;
    height: 100%;
    background-image: linear-gradient(white, #78c0e0);
`;


export const AppStyled = styled.div`
    display: flex;
    justify-content: center;
    overflow: hidden;
`;


export const AppStyledContent = styled.div`
    z-index: 10;
    position: absolute;
    height: 100%;
    overflow: auto;
    width: 100%;
    display: flex;
    justify-content: center;
`;