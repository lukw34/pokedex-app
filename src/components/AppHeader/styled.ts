import styled from 'styled-components';

export const AppHeaderStyled = styled.div`
    position: absolute;
    z-index: 15;
    height: 50px;
    top: 0;
    left: 0;
    width: 100vw;
    background: #c11616;
    justify-content: center;
    display: flex;
    align-items: center
`;

export const AppHeaderTitle = styled.div`
    flex: 1;
    font-weight: bold;
    color: white;
    font-size: 2rem;
    text-align: center;
`;

export const AppLogo = styled.img`
    width: 45px;
    height: 45px;
    margin: 5px;
`;