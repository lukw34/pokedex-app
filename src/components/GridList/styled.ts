import styled from 'styled-components';

export const PokemonGridListStyled = styled.div`
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    align-items: center;
    justify-content: center;
`;

export const GridListItem = styled.div`
    display: flex;
    align-items: center;
    justify-content: center;
`;