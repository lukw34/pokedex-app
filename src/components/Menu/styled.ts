import styled from 'styled-components';
import { Link } from 'react-router-dom';

export const MenuIconStyled = styled.img`
    width: 45px;
    height: 45px;
`;

export const MenuContainer = styled.div`
    z-index: 15;
    position: absolute;
    top: 50px;
    left: 0;
    background: #f3f1f1;
    height: 100vh;
    display: flex;
    flex-direction: column;
    align-items: center;
    box-shadow: 2px 0 18px rgba(0, 0, 0, 0.26);
`;


export const MenuItemLink = styled(Link)`
    border-bottom: 1px solid grey;
    width: 65px;
    height: 65px;
    display: flex;
    justify-content: center;
    align-items: center;
    transition: background .4s;
`;

export const MenuItemLinkContainer = styled.div<{ isActiveItem: boolean }>`
    background: ${props => props.isActiveItem ? '#d4242487' : '#f3f1f1'};
`;