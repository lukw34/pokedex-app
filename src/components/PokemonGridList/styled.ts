import styled from 'styled-components';

export const PokemonGridListStyled = styled.div`
    display: flex;
    flex-direction: row;
    flex-wrap: wrap;
    align-items: center;
    justify-content: center;
`;

export const PokemonGridListItem = styled.div`
    height: 250px;
    margin: 10px;
    flex: 0 0 20%;
    border: blue 2px solid;
    display: flex:
    flex-direction: row;
`;

export const PokemonName = styled.div`
    
`;