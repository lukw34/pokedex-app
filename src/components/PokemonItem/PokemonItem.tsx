import React, { useEffect, useState } from 'react';
import { PokemonDetails } from '../../types/pokemon';
import { EntityId } from '@reduxjs/toolkit';
import { useDispatch, useSelector } from 'react-redux';
import { fetchSinglePokemon, selectPokemonDetails } from '../../slices/pokedex.slice';
import { AppState } from '../../store';
import PokemonNoItem from './PokemonNoItem';
import { PokemonItemContainer, PokemonItemInnerContainer } from './styled';
import PokemonDetailsComp from './PokemonDetails';

interface PokemonItemProps {
    itemId: EntityId | undefined,
    itemNumber: number
}

const PokemonItem: React.FC<PokemonItemProps> = ({ itemId = '', itemNumber }) => {
    const [displayDetails, setDisplayDetails] = useState(false);
    const dispatch = useDispatch();
    const { url, ...props } = useSelector((state: AppState) => selectPokemonDetails(state, itemId) as PokemonDetails) || {};
    
    useEffect(() => {
        const isData = url === null;
        const itemNotFetched = url === undefined;
        let timeoutId: NodeJS.Timeout | null = null;
        if(!isData && !itemNotFetched) {
            dispatch(fetchSinglePokemon(url as string));
        } else if (!itemNotFetched) {
            timeoutId = setTimeout(() => {
                setDisplayDetails(true);
            }, 500);
        }

        return () => {
            if(timeoutId) {
                clearTimeout(timeoutId);
            }
        };

    }, [url]);

    return (
        <PokemonItemContainer displayDetails={displayDetails}>
            <PokemonItemInnerContainer displayDetails={displayDetails}>
                <PokemonNoItem/>
                <PokemonDetailsComp {...props} url={url} />
            </PokemonItemInnerContainer>
        </PokemonItemContainer>
    );
};
    


export default React.memo(PokemonItem);