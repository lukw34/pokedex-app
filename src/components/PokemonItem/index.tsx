import React from 'react';
import {Pokemon} from "../types/pokemon";

import { PokemonItemContainer, PokemonItemName, PokemonItemNoData, PokemonItemData } from './styled';

const PokemonItem: React.FC<Pokemon> = ({ name }) => {
    return (
        <PokemonItemContainer>
            <PokemonItemName>{name}</PokemonItemName>
            <PokemonItemNoData>#1</PokemonItemNoData>
            <PokemonItemData>Content</PokemonItemData>
        </PokemonItemContainer>
    )
};


export default PokemonItem;