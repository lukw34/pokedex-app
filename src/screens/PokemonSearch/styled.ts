import styled from 'styled-components';

export const PokemonSearchContainer = styled.div`
    display: flex;
    flex-direction: column;
    align-items: center;
    margin: 25px;
`;

export const PokemonInputContainer = styled.div`
    display: flex;
    flex-direction: row;
`;

export const PokemonInput = styled.input`
    flex: 1;
    width: 300px;
    height: 27px;
    border: 1px solid #c11616;
    text-align: center;
`;

export const PokemonSearchButton = styled.button`
    padding: 5px 20px;
    background: #c11616;
    border: 1px solid #c11616;
    color: white;
    font-weight: bold;
    font-size: 1rem;
`;

export const PokemonSearchPokemonItem = styled.div`
    height: 350px;
    margin: 15px;
    width: 250px;
    box-shadow: -5px 5px 8px 1px rgba(0,0,0,0.4);
    display: flex;
    flex-direction: row;
 `;
