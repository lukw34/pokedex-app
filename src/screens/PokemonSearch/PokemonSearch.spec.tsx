import React from 'react';
import { shallow } from 'enzyme';
import ReactRedux from 'react-redux';

import PokemonItemContainer from "../../components/PokemonItemContainer";
import PokemonItem from '../../components/PokemonItem';
import PokemoonSearch from '.';
import { PokemonInput, PokemonSearchButton } from './styled';

jest.mock('react-redux', () => ({
    useDispatch: jest.fn(),
    useSelector: jest.fn()
}));
const mockedReactRedux = ReactRedux as jest.Mocked<typeof ReactRedux>;

describe('<PokemonSearch />', () => {
    beforeEach(() => {
        jest.clearAllMocks();
    });
    it('should render render empty pokemon item on init', () => {
        const comp = shallow(<PokemoonSearch />);
        mockedReactRedux.useDispatch.mockReturnValueOnce(jest.fn());
        const { name } = comp.find(PokemonItemContainer).shallow().props();
        expect(name).toBe('???');
        expect(comp).toMatchSnapshot();
    });

    it('should render pokemon item if pokemon exist', () => {
        mockedReactRedux.useDispatch.mockReturnValueOnce(jest.fn());
        mockedReactRedux.useSelector.mockReturnValue({
            name: 'charmander',
            weight: 64,
            height: 45,
            image: 'image',
            types: ['fire'],
            stats: [],
            baseExperience: 20
        });
        const comp = shallow(<PokemoonSearch />);
        const itemContainer = comp.find(PokemonItemContainer).shallow();
        expect(itemContainer.props().name).toBe('charmander');
        expect(comp).toMatchSnapshot();
        expect(itemContainer.find(PokemonItem).shallow()).toMatchSnapshot();
    });

    it('handle input and trigger action on search click', () => {
        const dispatch = jest.fn();
        mockedReactRedux.useDispatch.mockReturnValueOnce(dispatch);
        const event = {target: { value: "1" }};
        const comp = shallow(<PokemoonSearch />);
        comp.find(PokemonInput).simulate('change', event);
        comp.find(PokemonSearchButton).simulate('click');
        expect(dispatch).toBeCalledTimes(1);
    });
});