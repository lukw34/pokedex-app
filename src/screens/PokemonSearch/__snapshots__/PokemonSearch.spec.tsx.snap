// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<PokemonSearch /> should render pokemon item if pokemon exist 1`] = `
<styled.div>
  <styled.div>
    <styled.input
      onChange={[Function]}
      value=""
    />
    <styled.button
      onClick={[Function]}
    >
      Search
    </styled.button>
  </styled.div>
  <styled.div>
    <PokemonItemContainer
      name=""
    />
  </styled.div>
</styled.div>
`;

exports[`<PokemonSearch /> should render pokemon item if pokemon exist 2`] = `
<styled.div>
  <styled.div>
    charmander
  </styled.div>
  <Styled(styled.div)
    isData={true}
  >
    #
    1
  </Styled(styled.div)>
  <Styled(styled.div)>
    <styled.img
      src="image"
    />
    <PokemonTypesComp
      id="charmander"
      types={
        Array [
          "fire",
        ]
      }
    />
    <styled.div>
      <styled.div
        title="Weight"
      >
        64
      </styled.div>
      <styled.div>
        20
      </styled.div>
      <styled.div
        title="Height"
      >
        45
      </styled.div>
    </styled.div>
    <PokemonStatsComp
      id="charmander"
      stats={Array []}
    />
  </Styled(styled.div)>
</styled.div>
`;

exports[`<PokemonSearch /> should render render empty pokemon item on init 1`] = `
<styled.div>
  <styled.div>
    <styled.input
      onChange={[Function]}
      value=""
    />
    <styled.button
      onClick={[Function]}
    >
      Search
    </styled.button>
  </styled.div>
  <styled.div>
    <PokemonItemContainer
      name=""
    />
  </styled.div>
</styled.div>
`;
